File: src/lib/scala/Slate.Server/src/main/scala/slate/server/utils/HttpHeaders.scala
/**
<slate_header>
  url: www.slatekit.com
  git: www.github.com/code-helix/slatekit
  org: www.codehelix.co
  author: Kishore Reddy
  copyright: 2016 CodeHelix Solutions Inc.
  license: refer to website and/or github
  about: A Scala utility library, tool-kit and server backend.
  mantra: Simplicity above all else
</slate_header>
  */

package slate.server.utils

import akka.http.scaladsl.server.RequestContext
import slate.common.{Strings, Inputs}

/**
  * abstraction layer over the akka http headers to support Protocol Independent APIs.
  * @param _ctx
  */
class HttpHeaders(private val _ctx:RequestContext) extends Inputs {


  /**
    * gets a string from the http headers
    * @param key
    * @return
    */
  override def getString(key: String) : String =
  {
    val header = _ctx.request.getHeader(key)
    if(header.isPresent) header.get.value else ""
  }


  /**
    * gets a string from the http headers if present or returns the default value
    * @param key
    * @param defaultVal
    * @return
    */
  override def getStringOrElse(key: String, defaultVal:String) : String =
  {
    val header = _ctx.request.getHeader(key)
    if(header.isPresent) header.get.value else defaultVal
  }


  /**
    * Gets a value from the header
    *
    * @param key
    * @return
    */
  override def getValue(key: String): AnyVal =
  {
    val text = getStringOrElse(key, "").toLowerCase
    val result =
      if(text == "true")
        true
      else if(text == "false")
        false
      else if(Strings.isInteger(text))
        text.toInt
      else if(Strings.isDouble(text))
        text.toDouble
      else if(text.length == 1 && text(0).isLetterOrDigit)
        text(0)
      else
        throw new IllegalArgumentException(s"key ${key} is not a value")
    result
  }


  /**
    * gets an object from the header data.
    *
    * @param key
    * @return
    */
  override def getObject(key: String): AnyRef =
  {
    val header = _ctx.request.getHeader(key)
    if(header.isPresent) header.get.value else ""
  }


  /**
    * Whether or not the key is present in the query string or in the post JSON post data.
    *
    * @param key
    * @return
    */
  override def containsKey(key: String): Boolean =
  {
    _ctx.request.getHeader(key).isPresent
  }


  override def size(): Int = {
    _ctx.request.headers.size
  }
}

File: src/lib/kotlin/slatekit-server/src/main/kotlin/slatekit/server/spark/HttpHeaders.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.server.spark

import slatekit.common.DateTime
import slatekit.common.InputFuncs
import slatekit.common.Inputs
import slatekit.common.encrypt.Encryptor
import spark.Request


data class HttpHeaders(val req: Request, val enc: Encryptor?) : Inputs {

    override fun get(key: String): Any? = getInternal(key)
    override fun getObject(key: String): Any? = getInternal(key)
    override fun containsKey(key: String): Boolean = req.headers().contains(key)
    override fun size(): Int = req.headers().size


    override fun getString(key: String): String = InputFuncs.decrypt(getInternalString(key).trim(), { it -> enc?.decrypt(it) ?: it })
    override fun getDate(key: String): DateTime = InputFuncs.convertDate(getInternalString(key).trim())
    override fun getBool(key: String): Boolean = getInternalString(key).trim().toBoolean()
    override fun getShort(key: String): Short = getInternalString(key).trim().toShort()
    override fun getInt(key: String): Int = getInternalString(key).trim().toInt()
    override fun getLong(key: String): Long = getInternalString(key).trim().toLong()
    override fun getDouble(key: String): Double = getInternalString(key).trim().toDouble()
    override fun getFloat(key: String): Float = getInternalString(key).trim().toFloat()


    fun getInternal(key: String): Any? {
        return if (containsKey(key)) {
            val value = req.headers(key)
            if (value != null && value is String) {
                value.trim()
            }
            else {
                value
            }
        }
        else {
            null
        }
    }


    fun getInternalString(key: String): String {
        return if (containsKey(key)) {
            val value = req.headers(key)
            if (value != null && value is String) {
                value.trim()
            }
            else {
                value
            }
        }
        else {
            ""
        }
    }
}